File: Forth_Interpreter/OperatorDoLoop.cpp
#include "OperatorDoLoop.hpp"

#include <sstream>
#include <vector>

bool OperatorDoLoop::Run () {
    GeneralDataForExecutors.CommandText.erase(GeneralDataForExecutors.CommandText.begin(), GeneralDataForExecutors.CommandText.begin() + 3);
    GeneralDataForExecutors.CommandText.erase(GeneralDataForExecutors.CommandText.end() - 7, GeneralDataForExecutors.CommandText.end());
    
    (*GeneralDataForExecutors.CommandsForSecondProcessing).clear();

    if (GeneralDataForExecutors.Stack->empty()) {
        throw Exception_EmptyStack();    
    }

    int i = GeneralDataForExecutors.Stack->top();
    GeneralDataForExecutors.Stack->pop();

    if (GeneralDataForExecutors.Stack->empty()) {
        throw Exception_EmptyStack();    
    }

    int N = GeneralDataForExecutors.Stack->top();
    GeneralDataForExecutors.Stack->pop();

    std::istringstream ist (GeneralDataForExecutors.CommandText);
    std::string tmp;
    std::vector <int> indexesI;
    int balanceDoLoop = 1;
    bool flagDoubleQuotes = false;
    int indexNow = -1;

    while (ist >> tmp) {
        indexNow += tmp.length();

        if (tmp == ".\"") {
            flagDoubleQuotes = true;
        }
        else if (tmp[tmp.length() - 1] == '"') {
            flagDoubleQuotes = false;
        }

        if (!flagDoubleQuotes) {
            if (tmp == "do") {
                balanceDoLoop++;
            }
            else if (tmp == "loop") {
                balanceDoLoop--;
            }
            else if (balanceDoLoop == 1 && tmp == "i") {
                indexesI.push_back(indexNow);
            }
        }

        indexNow++;
    }

    std::string result;
    std::string pattern = GeneralDataForExecutors.CommandText;

    for (; i < N; ++i) {
        GeneralDataForExecutors.CommandText = pattern;
        std::ostringstream ost;
        ost << i;

        for (int j = indexesI.size() - 1; j >= 0 ; --j) {
            GeneralDataForExecutors.CommandText.replace(indexesI[j], 1, ost.str());
        }

        result += GeneralDataForExecutors.CommandText + " ";
    }

    if (result.length() > 0) {
        result.pop_back();
    }

    *GeneralDataForExecutors.CommandsForSecondProcessing = result;

    return false;
}

namespace ForthFactoryRegistrations {
    FactoryRegistrationCommands<OperatorDoLoop> Command_do ("do");
}

File: Forth_Interpreter_On_Java/src/main/java/ru/nsu/yurakov/Forth/OperatorDoLoop/OperatorDoLoop.java
package ru.nsu.yurakov.Forth.OperatorDoLoop;

import java.util.EmptyStackException;
import java.util.ArrayList;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ru.nsu.yurakov.Forth.ForthCommands.ForthCommands;
import ru.nsu.yurakov.Forth.ForthRuntimeException.ForthRuntimeException;

/**
 * Unwinds the loop and substitutes the desired values instead of i.<br>
 * And sends the result for secondary processing
 */
public class OperatorDoLoop extends ForthCommands{

    @Override
    public boolean Run() throws ForthRuntimeException {
        LOGGER.info("Was launched " + this.getClass().getName() + " <" + GeneralDataForExecutors.CommandText + ">");
        try {
            String command = GeneralDataForExecutors.CommandText;
            command = command.substring(3, command.length() - 7);
            String[] commands = command.split(" ");
            String pattern = command;

            int i = GeneralDataForExecutors.stack.peek();
            GeneralDataForExecutors.stack.pop();

            int N = GeneralDataForExecutors.stack.peek();
            GeneralDataForExecutors.stack.pop();

            ArrayList<Integer> indexesI = new ArrayList<>();
            int balanceDoLoop = 1;
            boolean flagDoubleQuotes = false;
            int indexNow = -1;

            for (String s : commands) {
                command = s;
                indexNow += command.length();

                if (command.equals(".\"")) {
                    flagDoubleQuotes = true;
                } else if (command.charAt(command.length() - 1) == '"') {
                    flagDoubleQuotes = false;
                }

                if (!flagDoubleQuotes) {
                    if (command.equals("do")) {
                        balanceDoLoop++;
                    } else if (command.equals("loop")) {
                        balanceDoLoop--;
                    } else if (balanceDoLoop == 1 && command.equals("i")) {
                        indexesI.add(indexNow);
                    }
                }

                ++indexNow;
            }

            String result = "";

            for (; i < N; ++i) {
                String newCommand = pattern;

                for (int j = indexesI.size() - 1; j >= 0 ; --j) {
                    String num = Integer.toString(i);
                    newCommand = newCommand.substring(0, indexesI.get(j)) 
                                 + num 
                                 + newCommand.substring(indexesI.get(j) + num.length());
                }

                result += newCommand + " ";
            }

            if (result.length() > 0) {
                result = result.substring(0, result.length() - 1);
            }

            GeneralDataForExecutors.CommandsForSecondProcessing = result;
        } catch (EmptyStackException e) {
            throw new ForthRuntimeException("Stack is empty");
        }

        return false;
    }

    private static final Logger LOGGER = LogManager.getLogger(OperatorDoLoop.class);
}
